### ISSUE TITLE
Tests ordering

### ISSUE THREAD

**Issue by samaursa** [catchorg/Catch2#60]
This is a features request. It is of low priority but will be nice to have. Maybe the feature already exists?

I have tests in different source (.cpp) files. Each cpp file has a group of related tests. Sometimes, one group of tests relies on objects that are also tested but in another group. I would like to order them such that the group of tests that relies on some objects is tested AFTER the group of tests that test the objects in the first place.

That may have been confusing so here is an example:

GROUP1 - tests for sorting on containers (containerSortTests.cpp)
GROUP2 - tests for containers to make sure they work properly (containerTests.cpp) 

I would like Group2 tests to be tested before Group1. If Group2 tests fail, then Group1 will most likely fail regardless. The 'ordering' of the failure will tell me to fix the containers first.

**Comment by acidtonic** [comment]
I too desire to have this functionality. I think it's as simple as allowing a test definition to contain a third field of pre-requisites.

Instead of...
TEST_CASE("sorting", "test basic sort routines");
TEST_CASE("container", "Tests the base containers");
TEST_CASE("container_sorting", "Tests the containers sorting");

we could instead rewrite that as
TEST_CASE("sorting", "test basic sort routines");
TEST_CASE("container", "Tests the base containers");
TEST_CASE("container_sorting", "Tests the containers sorting", "{container,sorting}");

Where the added third parameter is simply the name of test cases that must pass without failure before this testcase will run.... Thus container_sorting tests would only be ran if the test label "container" and "sorting" had no failures.

I too would be interested in implementing this feature if it's not a priority.

**Comment by KazukiCP** [comment]
I was also looking for this functionality. 
Since it's not implemented I though I would share a work-around I found using a custom main function.

Assuming the following tests (I took the example of acidtonic)
TEST_CASE("sorting", "test basic sort routines");
TEST_CASE("container", "Tests the base containers");
TEST_CASE("container_sorting", "Tests the containers sorting");

The main could be written as follow

    #define CATCH_CONFIG_RUNNER
    #include <Catch2/catch.hpp>

    int main(int argc, char** argv)
    {
        Catch::Session s;
        int init_code = s.applyCommandLine(argc, argv);
        if (init_code != 0)
        {
            return init_code;
        }
    
        auto config_data = s.configData();
        int num_failed = 0;
    
        for (std::string_view tag : { "[container]", "[sorting]" })
        {
            config_data.testsOrTags = { std::string(tag) };
            s.useConfigData(config_data);
            num_failed = s.run();
        }
    
        if (!num_failed)
        {
            for (std::string_view tag : { "[container_sorting]" })
            {
                config_data.testsOrTags = { std::string(tag) };
                s.useConfigData(config_data);
                num_failed = s.run();
            }
        }

        return num_failed;
    }

**Comment by AlexisWilke** [comment]
I was just wondering whether the code supported such a feature...

My current solution is to use a mix of the above:

* I create groups using tags in each set of test cases (as shown by **KazukiCP**)

      // test cases are assigned different tags as follow
      TEST_CASE("some_name", "[container]")

* As a result, I can create a single binary to run all the tests (as requested by **samaursa**)
* I run a specific set of tests using the tags through a bash script, that bash script defines the order (as mentioned by **philsquared**)

      // running the tests
      #!/bin/bash -e
      unittest [container] &
      unittest [sorting] &
      wait
      unittest [container_sorting]

  The `-e` option tells bash to exit on errors.

  The `&` runs those tests in the background and thus I can run two `unittest` instances in parallel.

  The `wait` is a synchronization point which blocks until the previous `unittest` instances are done.

  Obviously, the order within a group is _undefined_. As mentioned above, defining sections within a test is a way to run your tests in a given order. However, that could mean a gigantic test case and that's probably not a good idea (really hard to maintain).
